### ISSUE TITLE
False positive when assigning to Final instance attribute in __init__ and adding generics

### ISSUE THREAD

**Issue by AndreuCodina** [astral-sh/ty#1888]
### Summary

Similar to the closed https://github.com/astral-sh/ty/issues/1409, but when adding generics.

```python
import asyncio
from typing import Final


class AsyncConcurrentDictionary[TKey, TValue]:
    _dict: Final[dict[TKey, TValue]]
    _lock: Final[asyncio.Lock]

    def __init__(self) -> None:
        self._dict = {}  # error[invalid-assignment]: Cannot assign to final attribute `_dict` on type `Self@__init__`
        self._lock = asyncio.Lock()  # error[invalid-assignment]: Cannot assign to final attribute `_lock` on type `Self@__init__`
```

https://play.ty.dev/2cb2f484-013d-4da2-a7f4-32a0144f5273

After some experiments, I figured out that if you remove the generics or repeat the type hints in `__init__`, the errors disappear, i.e.:

```python
import asyncio
from typing import Final


class AsyncConcurrentDictionary:
    _dict: Final[dict[int, int]]
    _lock: Final[asyncio.Lock]

    def __init__(self) -> None:
        self._dict = {}
        self._lock = asyncio.Lock()

# or

class AsyncConcurrentDictionary[TKey, TValue]:
    _dict: Final[dict[TKey, TValue]]
    _lock: Final[asyncio.Lock]

    def __init__(self) -> None:
        self._dict: dict[TKey, TValue] = {}
        self._lock: asyncio.Lock = asyncio.Lock()
```

### Version

0.0.1a34

**Comment by AndreuCodina** [comment]
Hi @charliermarsh

You marked the issue as completed, but I still have errors with that code.

```python
import asyncio
from typing import Final


class AsyncConcurrentDictionary[TKey, TValue]:
    _dict: Final[dict[TKey, TValue]]
    _lock: Final[asyncio.Lock]

    def __init__(self) -> None:
        self._dict = {}
        self._lock = asyncio.Lock()
```

I tested it with a new project:

```bash
uv init --lib show-error
cd show-error
uv add --dev ty
```

**Comment by sharkdp** [comment]
The change has been merged, but it has not been released yet. The playground runs the latest version of ty from the main branch and you can see that the error is resolved there: [Playground](https://play.ty.dev/f2885230-d9cd-40a1-989c-a07461a0c2cf). So this should be fixed in the next release.
